Test of Python <--> JS interface

giangiac opened the pull request.

giangiac committed:
Add 4th tab to ggg_front

giangiac committed:
[WIP] Test of interaction with Python backend

giangiac committed:
[WIP] Test interface with backend

giangiac commented on ggg_frontend/src/templates/index.js:
```diff
@@ -0,0 +1,224 @@
+
+/////////////////////////////////////////////////////////////
+
+	const testButton = button('test of API to interface with API', gridSize, 1*gridSize, 'orange', 'black');
```
Here are the buttons to 'get' and 'post' using the interface.

Gistbatch commented:
Hi Gian, Here is the good news: it's already working correctly!
[Screencast from 20.06.2024 13:46:47.webm](https://github.com/giangiac/tqec/assets/18579069/bdd53392-16f4-4f84-b5bf-63bce465e99d)

What are you seeing:
- First, I start the Flask server in the backend `python web/main.py`
- Then I press the get button in the interface
- In the browser console, you can see the status `200`, which means OK, and the value at the corresponding address `localhost/example`, which is `42`.
- In the terminal, you see the corresponding action: `Sending: {'name': 'example', 'value': 42}
`
- After I press the post button which is supposed to send a json object `{name: 'post_example', value: '3.14'}` to the same location. 
- In the browser console, you again retrieve the status `200 OK`.
- On the terminal, you see the corresponding action: `Received: {'name': 'post_example', 'value': '3.14'}
`

If you have questions about what the code does, please let me know.

inmzhang committed:
Implement SketchUp-exported dae file parsing with `pycollada` (#253)

Parse SketchUp-exported dae file with `pycollada`.
---------

Co-authored-by: Adrien Suau <nelimee.dev+github@protonmail.com>

nelimee committed:
Implement `TemplateGrid` (#245)

This PR implements a generic `TemplateGrid` that can be used to generate
a grid of templates representing logical qubits (for example).

Closes #244

nelimee committed:
Use scalable variant in templates (#254)

This PR changes how scaling is performed in `tqec`.

This fix #241.

**Before**

Before that PR, the `Dimension` class was storing a `LinearFunction`
describing how the `Dimension` was supposed to scale and a `value`
representing the current value of the `Dimension`.

The `Template` instance were taking the `Dimension` instances as input,
and were using them to scale internally, but were only outputting
integers resulting from the "instantiation" of such `Dimension`
instances.

**In this PR**

This PR completely removes the `Dimension` class. `Template` instances
are now responsible for storing the scaling parameter `k` along with the
`LinearFunction` instances needed to scale.

The `Template` class now requires the implementation of a new
`scalable_shape` property that returns the shape of the `Template`
instance as a fully scalable structure. It basically returns the full
information (the full function) instead of returning its value on one
specific point (before). For that, `ComposedTemplate` needed a
`PiecewiseLinearFunction` class due to the use of `min`/`max` in the
computation of the methods needed for `scalable_shape`.

As side (but important) notes:
- the `ShiftedTemplate` and `StackedTemplate` classes are removed from
the code base, because they are not currently used and are more classes
that would need maintenance.
- all the JSON-related methods are also removed, due to their
replacement in #240.

---------

Co-authored-by: Yiming Zhang <61700160+inmzhang@users.noreply.github.com>

giangiac commented:
> Hi Gian, Here is the good news: it's already working correctly! [Screencast from 20.06.2024 13:46:47.webm](https://github.com/giangiac/tqec/assets/18579069/bdd53392-16f4-4f84-b5bf-63bce465e99d)
> 
> What are you seeing:
> 
> * First, I start the Flask server in the backend `python web/main.py`
> * Then I press the get button in the interface
> * In the browser console, you can see the status `200`, which means OK, and the value at the corresponding address `localhost/example`, which is `42`.
> * In the terminal, you see the corresponding action: `Sending: {'name': 'example', 'value': 42} `
> * After I press the post button which is supposed to send a json object `{name: 'post_example', value: '3.14'}` to the same location.
> * In the browser console, you again retrieve the status `200 OK`.
> * On the terminal, you see the corresponding action: `Received: {'name': 'post_example', 'value': '3.14'} `
> 
> If you have questions about what the code does, please let me know.

Thanks a lot for the breakdown of the actions!

nelimee committed:
Fix grid.py and grid_test.py (#255)

This fixes the last merge that has been performed too quickly.

giangiac committed:
Make the guideline appear when GET (from backend) button is clicked

giangiac committed:
Minor edits

giangiac committed:
Specialize the behavior of dragging a PlaquetteType depending on the workspace

giangiac committed:
Update the data from frntend to backend

giangiac committed:
Simplify code by using common utility functions

giangiac committed:
Rename the folder adding 'tab_' when appropriate

giangiac committed:
Communicate default json files for templates

giangiac committed:
Add position of top-left corner to plaquettes

giangiac committed:
Add labels 1 and 2 to the template cells

nelimee committed:
Add detector automation (#249)

This is a large PR with a lot of new code and a lot of code from #227. I
will summarize everything here, even things that have already been
summarized in #227.

There are still minor issues to solve, but most of the PR should be
ready to review. The minor issues are listed at the end of this (very
long, sorry) message.

### General considerations

This PR implements the [proposal by
Austin](https://github.com/QCHackers/tqec/issues/199#issuecomment-2015637349).

It does so fully in `stim`, not using `cirq`, and not using most of the
already implemented functions in the TQEC package. This is a deliberate
choice as this PR implements a feature that:
1. would not highly benefit from a `cirq` intermediary representation,
2. might very well be outsourced into its own package and distributed
separately from the TQEC package to the community eventually.

A lot of code has been taken from #227, either as-is or as a base that
has been changed when developing new features.

The code is organized into 4 main problems: the 3 described in [this
comment](https://github.com/QCHackers/tqec/pull/227#issuecomment-2155060762)
and a fourth one that is compiling back the fragments and obtained
detectors into a `stim.Circuit` instance. The following sections
highlight the main points used to solve each of these 4 problems.

To illustrate, I will use a simple repetition code of distance 3 and
using 5 rounds obtained with:
```py
import stim
circuit = stim.Circuit.generated("repetition_code:memory", distance=3, rounds=5)
# Detail: insert QUBIT_COORDS into the circuit
```
giving the circuit
```
QUBIT_COORDS(0) 0
QUBIT_COORDS(1) 1
QUBIT_COORDS(2) 2
QUBIT_COORDS(3) 3
QUBIT_COORDS(4) 4
R 0 1 2 3 4
TICK
CX 0 1 2 3
TICK
CX 2 1 4 3
TICK
MR 1 3
DETECTOR(1, 0) rec[-2]
DETECTOR(3, 0) rec[-1]
REPEAT 4 {
    TICK
    CX 0 1 2 3
    TICK
    CX 2 1 4 3
    TICK
    MR 1 3
    SHIFT_COORDS(0, 1)
    DETECTOR(1, 0) rec[-2] rec[-4]
    DETECTOR(3, 0) rec[-1] rec[-3]
}
M 0 2 4
DETECTOR(1, 1) rec[-2] rec[-3] rec[-5]
DETECTOR(3, 1) rec[-1] rec[-2] rec[-4]
OBSERVABLE_INCLUDE(0) rec[-1]
```

![image](https://github.com/QCHackers/tqec/assets/12374487/110aa35f-2b1d-4a8e-80dd-0f24e3e17e2e)


### Sub-problem 1: Splitting the input circuit

The `Fragment` approach from #227 has been a little bit changed, but the
overall idea stayed.

The provided circuit is split into "fragments". A fragment is, loosely
speaking, equivalent to the circuit needed to perform one QEC round. In
more details, a fragment can either be an instance of `Fragment` or
`FragmentLoop`.

`FragmentLoop` is the simplest one: it is a structure that represents
`REPEAT` blocks from `stim` by storing the loop body (as a list of
instances of either `Fragment` or `FragmentLoop`) and a number of
repetitions.

`Fragment` represents the "atomic" instances, that cannot be subdivided
into sub-fragments. A `Fragment` instance stores a portion of a
`stim.Circuit` that follow the scheme
```
(resets) -> (computations) -> measurements
```
with `resets` and `computations` being optional but `measurements` being
a requirement.

This restriction will make more sense in step 2. It basically helps
making all the code simpler, at the expense of one more point that need
to be handled: combined measurement/reset gates.

#### Combined measurement/reset gates

In the circuit above, the gate `MR` implements a measurement directly
followed by a reset operation.

These gates are convenient, but they break the pre-condition of
`Fragment` exposed above. In order to restore that pre-condition, input
`stim.Circuit` are transformed by:

1. splitting combined gates into their non-combined equivalents. For
example, `MR -> M and then R` or `MRX -> MX and then RX`.
2. replacing each portion of the circuit that match the following
template
    ```
    R [qubits];
    [Annotations / Noisy gates]
    REPEAT [N] {
        [Computation gates (no reset, no measurement)]
        M [qubits]
        R [qubits]
        [Annotations / Noisy gates]
    }
    ```
    by the functionally equivalent circuit
    ```
    REPEAT [N] {
        R [qubits]
        [Annotations / Noisy gates]
        [Computation gates (no reset, no measurement)]
        M [qubits]
    }
    ```
the only difference being that the last reset of the loop is never
applied, which might be an issue in some cases, so we might in the
future replace the above circuit by
    ```
    REPEAT [N] {
        R [qubits]
        [Annotations / Noisy gates]
        [Computation gates (no reset, no measurement)]
        M [qubits]
    }
    R [qubits];
    [Annotations / Noisy gates]
    ```
    for exact equivalence.

So in the end, users can call `split_stim_circuit_into_fragments` in
order to split any valid `stim.Circuit` into a list of fragments:
```py
import stim
from tqec.circuit.detectors.utils import (
    split_combined_measurement_reset,
    reorder_resets,
)

circuit = reorder_resets(
    split_combined_measurement_reset(
        stim.Circuit.generated("repetition_code:memory", distance=3, rounds=5)
    )
)
```
resulting in
```
R 0 1 2 3 4
TICK
CX 0 1 2 3
TICK
CX 2 1 4 3
TICK
M 1 3
DETECTOR(1, 0) rec[-2]
DETECTOR(3, 0) rec[-1]
TICK
REPEAT 4 {
    R 1 3
    TICK
    CX 0 1 2 3
    TICK
    CX 2 1 4 3
    TICK
    M 1 3
    SHIFT_COORDS(0, 1)
    DETECTOR(1, 0) rec[-2] rec[-4]
    DETECTOR(3, 0) rec[-1] rec[-3]
    TICK
}
M 0 2 4
DETECTOR(1, 1) rec[-2] rec[-3] rec[-5]
DETECTOR(3, 1) rec[-1] rec[-2] rec[-4]
OBSERVABLE_INCLUDE(0) rec[-1]
```

![circuit_in_fragments](https://github.com/QCHackers/tqec/assets/12374487/c8b99afd-d0de-4b71-ab71-9716002a816e)
with `Fragment` instances in blue boxes and `FragmentLoop` instances in
green boxes and represented in code by:
```py
[
    Fragment(
        circuit=stim.Circuit("""
            R 0 1 2 3 4
            TICK
            CX 0 1 2 3
            TICK
            CX 2 1 4 3
            TICK
            M 1 3
            DETECTOR(1, 0) rec[-2]
            DETECTOR(3, 0) rec[-1]
            TICK
        """)
    ),
    FragmentLoop(
        repetitions=4,
        fragments=[
            Fragment(
                circuit=stim.Circuit("""
                    R 1 3
                    TICK
                    CX 0 1 2 3
                    TICK
                    CX 2 1 4 3
                    TICK
                    M 1 3
                    SHIFT_COORDS(0, 1)
                    DETECTOR(1, 0) rec[-2] rec[-4]
                    DETECTOR(3, 0) rec[-1] rec[-3]
                    TICK
                """)
            )
        ],
    ),
    Fragment(
        circuit=stim.Circuit("""
            M 0 2 4
            DETECTOR(1, 1) rec[-2] rec[-3] rec[-5]
            DETECTOR(3, 1) rec[-1] rec[-2] rec[-4]
            OBSERVABLE_INCLUDE(0) rec[-1]
        """)
    ),
]
```

### Sub-problem 2: Pre-computing important values

Detectors are determined thanks to the knowledge of stabilizer
propagation. This means that, in order to hope to automatically compute
detectors, we need to know what are the stabilizers that are created,
destructed and propagated in each fragment.

This is where the restriction on `Fragment` from step 1 makes all its
sense. As reset gates can only be at the beginning of a `Fragment` and
measurements gates can only be at the end of it, we start by computing
two quantities **locally** on each `Fragment`:

- `creation` stabilizers are stabilizers that are "created" by reset
gates and propagate forward in the circuit represented by the
`Fragment`, until just before the measurements where its propagation is
stopped and the resulting propagated stabilizer is stored into a
`BoundaryStabilizer` instance, along with the collapsing operations
(here measurements) that the stabilizer will have to go through in the
next timestep.
- similarly, `destruction` stabilizers are stabilizers that are
"destructed" (or ended) by measurements. They are computed by
propagating the stabilizer "destructed" by each measurement backward,
and storing the resulting propagated stabilizer just before reaching the
resets (or the beginning of the `Fragment` if there is no reset),
storing that information into a `BoundaryStabilizer` instance, along
with the collapsing operations (here resets) that the stabilizer will
have to go through in the previous timestep. Another way of viewing
`destruction` stabilizers computation is the following: take the
circuit, revert it (**not** inverting, simply applying the same gates,
but in reverse order), change measurements in resets, resets in
measurements, and compute the `creation` stabilizers described above.
These are the `destruction` stabilizers of the initial circuit.

Because we intentionally restrict ourselves to detectors that only
involve measurements from the current and previous QEC round (loosely
speaking, the current and previous `Fragment`), we do not have to
compute any "propagated" stabilizer, i.e. a structure that would compute
and store, for each `Fragment` a map of input stabilizers and the
corresponding output stabilizer.

All this `creation`/`destruction` stabilizers are called flows, and
stored in `FragmentFlows` instances.

For `FragmentLoop`, it is considered as an atomic operation and its
`creation` (resp. `destruction`) flows are the `creation` (resp.
`destruction`) flows of the last (resp. first) fragment of its body.
These are stored in `FragmentLoopFlows` instances.

For example, the flows computed on the example of fragments above are:
```py
[
    FragmentFlows(
        creation=[
            BoundaryStabilizers(
                stabilizer=PauliString(qubits={0: "Z"}),
                collapsing_operations=[],
                involved_measurements=[],
            ),
            BoundaryStabilizers(
                stabilizer=PauliString(qubits={0: "Z", 1: "Z", 2: "Z"}),
                collapsing_operations=[PauliString(qubits={1: "Z"})],
                involved_measurements=[
                    RelativeMeasurementLocation(offset=-2, qubit_index=1)
                ],
            ),
            BoundaryStabilizers(
                stabilizer=PauliString(qubits={2: "Z"}),
                collapsing_operations=[],
                involved_measurements=[],
            ),
            BoundaryStabilizers(
                stabilizer=PauliString(qubits={2: "Z", 3: "Z", 4: "Z"}),
                collapsing_operations=[PauliString(qubits={3: "Z"})],
                involved_measurements=[
                    RelativeMeasurementLocation(offset=-1, qubit_index=3)
                ],
            ),
            BoundaryStabilizers(
                stabilizer=PauliString(qubits={4: "Z"}),
                collapsing_operations=[],
                involved_measurements=[],
            ),
        ],
        destruction=[
            BoundaryStabilizers(
                stabilizer=PauliString(qubits={0: "Z", 1: "Z", 2: "Z"}),
                collapsing_operations=[
                    PauliString(qubits={0: "Z"}),
                    PauliString(qubits={1: "Z"}),
                    PauliString(qubits={2: "Z"}),
                ],
                involved_measurements=[
                    RelativeMeasurementLocation(offset=-2, qubit_index=1)
                ],
            ),
            BoundaryStabilizers(
                stabilizer=PauliString(qubits={2: "Z", 3: "Z", 4: "Z"}),
                collapsing_operations=[
                    PauliString(qubits={2: "Z"}),
                    PauliString(qubits={3: "Z"}),
                    PauliString(qubits={4: "Z"}),
                ],
                involved_measurements=[
                    RelativeMeasurementLocation(offset=-1, qubit_index=3)
                ],
            ),
        ],
        total_number_of_measurements=2,
    ),
    FragmentLoopFlows(
        fragment_flows=[
            FragmentFlows(
                creation=[
                    BoundaryStabilizers(
                        stabilizer=PauliString(qubits={0: "Z", 1: "Z", 2: "Z"}),
                        collapsing_operations=[PauliString(qubits={1: "Z"})],
                        involved_measurements=[
                            RelativeMeasurementLocation(offset=-2, qubit_index=1)
                        ],
                    ),
                    BoundaryStabilizers(
                        stabilizer=PauliString(qubits={2: "Z", 3: "Z", 4: "Z"}),
                        collapsing_operations=[PauliString(qubits={3: "Z"})],
                        involved_measurements=[
                            RelativeMeasurementLocation(offset=-1, qubit_index=3)
                        ],
                    ),
                ],
                destruction=[
                    BoundaryStabilizers(
                        stabilizer=PauliString(qubits={0: "Z", 1: "Z", 2: "Z"}),
                        collapsing_operations=[PauliString(qubits={1: "Z"})],
                        involved_measurements=[
                            RelativeMeasurementLocation(offset=-2, qubit_index=1)
                        ],
                    ),
                    BoundaryStabilizers(
                        stabilizer=PauliString(qubits={2: "Z", 3: "Z", 4: "Z"}),
                        collapsing_operations=[PauliString(qubits={3: "Z"})],
                        involved_measurements=[
                            RelativeMeasurementLocation(offset=-1, qubit_index=3)
                        ],
                    ),
                ],
                total_number_of_measurements=2,
            )
        ],
        repeat=4,
    ),
    FragmentFlows(
        creation=[],
        destruction=[
            BoundaryStabilizers(
                stabilizer=PauliString(qubits={0: "Z"}),
                collapsing_operations=[],
                involved_measurements=[
                    RelativeMeasurementLocation(offset=-3, qubit_index=0)
                ],
            ),
            BoundaryStabilizers(
                stabilizer=PauliString(qubits={2: "Z"}),
                collapsing_operations=[],
                involved_measurements=[
                    RelativeMeasurementLocation(offset=-2, qubit_index=2)
                ],
            ),
            BoundaryStabilizers(
                stabilizer=PauliString(qubits={4: "Z"}),
                collapsing_operations=[],
                involved_measurements=[
                    RelativeMeasurementLocation(offset=-1, qubit_index=4)
                ],
            ),
        ],
        total_number_of_measurements=3,
    ),
]
```

### Sub-problem 3: Matching stabilizer flows

This is the main computational step. This PR mostly follows [Austin's
proposal](https://github.com/QCHackers/tqec/issues/199#issuecomment-2015637349).

The only real change from #227 is that the function to find a cover has
been radically changed.
The initial implementation was performing an exhaustive search on all
the possible combinations of stabilizers. First, there was an initial
filtering on candidates for the cover [that was not entirely
correct](https://github.com/QCHackers/tqec/pull/227#issuecomment-2155060762).
Removing this filtering increased the number of stabilizers to consider,
and blew up the computational cost of finding the cover.
I tried to filter again stabilizers, but this time on the assumption
that most detectors will be spatially local (which is false for a
variety of QEC codes, but true for our initial focus). This locality
assumption improved the runtime, but was still not enough, some detector
computations taking tens of seconds.
So I ended up rephrasing the cover problem as a SAT problem, solved by a
specific solver that accept SAT formulations with XOR clauses instead of
OR clauses: [`pycryptosat`](https://pypi.org/project/pycryptosat/)
accessed through the
[`pysat`](https://pysathq.github.io/docs/html/api/solvers.html#module-description)
interface. It turns out that this rephrasing allowed to solve any (even
non-filtered with more than 40 stabilizers considered for the cover)
cover problems in a very small time (not user-noticeable anymore, I have
not benchmarked it properly).

### Sub-problem 4: re-building the circuit

Re-building the circuit from the fragments is super easy:
```py
def compile_fragments_to_circuit(
    fragments: list[Fragment | FragmentLoop],
) -> stim.Circuit:
    circuit = stim.Circuit()

    for fragment in fragments:
        if isinstance(fragment, Fragment):
            circuit += fragment.circuit
        else:  # isinstance(fragment, FragmentLoop):
            loop_body = compile_fragments_to_circuit(fragment.fragments)
            circuit += loop_body * fragment.repetitions
    return circuit
```
Computing and inserting detectors is not really more difficult:
```py
def compile_fragments_to_circuit_with_detectors(
    fragments: list[Fragment | FragmentLoop],
    qubit_coords_map: dict[int, tuple[float, ...]],
) -> stim.Circuit:
    flows = build_flows_from_fragments(fragments)
    detectors = match_detectors_from_flows_shallow(flows, qubit_coords_map)

    circuit = stim.Circuit()

    for fragment, detectors in zip(fragments, detectors):
        detectors_circuit = _detectors_to_circuit(detectors, [0.0])
        if isinstance(fragment, Fragment):
            circuit += (
                _remove_last_tick_instruction(fragment.circuit)
                + detectors_circuit
                + _tick_circuit()
            )
        else:  # isinstance(fragment, FragmentLoop):
            shift_circuit = _shift_time_instruction(len(detectors[0].coords))
            loop_body = compile_fragments_to_circuit_with_detectors(
                fragment.fragments, qubit_coords_map
            )
            circuit += (
                _remove_last_tick_instruction(loop_body)
                + shift_circuit
                + detectors_circuit
                + _tick_circuit()
            ) * fragment.repetitions
    return circuit
```

## Issues still present

### Missing tests

The code is currently quite well documented, I tried to document
thoroughly each and every function / class, but there is no test at all.
Most of the tests from #227 can likely be adapted, but way more tests
will have to be implemented for all the functions / classes introduced
in this PR.

### Temporal coordinate of detectors

Currently, detectors only have `0.0` in their temporal coordinate. I am
trying to find a nice way to devise the temporal coordinate of each
detector. I will definitely have a look at how this is done in #227 in
the next few days and see if I can use the same method here.

---------

Co-authored-by: Yiming Zhang <61700160+inmzhang@users.noreply.github.com>
Co-authored-by: Philipp Seitz <philipp.seitz@tum.de>

giangiac committed:
Fill template

giangiac committed:
Adopt dictionary {x, y} for guide constants

nelimee committed:
Cleanup a few places in the repository (#264)

This PR cleans up:
- a folder containing one notebook that is now outdated,
- the `specification/schemas` folder, because it is outdated and #240
will eventually provide that automatically,
- splits `tqec.enums` into `tqec.template.enums` and
`tqec.plaquette.enums` to clean up a little bit the main `tqec`
namespace.

giangiac committed:
Add dropdown menu to select the tempalte

giangiac committed:
[WIP] Use instantiation field to get the template shape

giangiac committed:
Improve drawing the guides from the template instantiation

giangiac committed:
Use a single json for the template library

giangiac committed:
Get alternative frontend back in the codebase (#250)

giangiac committed:
Solve merge conflicts

giangiac committed:
Update info tab; add example of template with 3 labels

giangiac committed:
Fill cells with labels different from 1 and 2

giangiac committed:
Extend message returned by frontend; simplify code

The message relates to the filled template in tab 4

giangiac commented:
This was an utility PR to facilitate the comparison between two development snapshots which differed only on the relevant feature.

The issue has been solved and this PR has outlived its utility.
